### ISSUE TITLE
Add a listener for asset events (not just assets)

### ISSUE THREAD

**Issue by jrderuiter** [apache/airflow#42254]
### Description

I'd like to add a new dataset listener that allows me to listen for new dataset events. This would be similar to the existing `on_dataset_changed` listener, but would pass the underlying `DatasetEvent` instead of the `Dataset` reference:

```
@hookspec
def on_dataset_event_created(
    event : DatasetEvent,
):
    """Execute when dataset a dataset event is created."""
```

Alternatively, we could expand the existing `on_dataset_changed` handler to add the dataset event as an argument to the listener. This would however not be backwards compatible with existing listener implementations.

### Use case/motivation

Currently Airflow provides the `on_dataset_changed` [listeners](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/listeners.html#dataset-events) as a hook for listening to changes on datasets. However, this listener only passes the dataset object to the listener and not the `DatasetEvent` corresponding to the change:

```
@hookspec
def on_dataset_changed(
    dataset: Dataset,
):
    """Execute when dataset change is registered."""
```

For a use case that I'm working on I'd also like to have access to the dataset event to determine whether that dataset event came from the API or not (which you can check on the event using `event.extra["from_rest_api"]`.

This is not possible with the current listeners.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by uranusjr** [comment]
Makes sense. Would you mind submitting a PR? You should start with the DatasetManager class; see how `on_dataset_changed` is notified for an example. You also need to add the hook signature to `airflow/listeners/spec/dataset.py`.

**Comment by uranusjr** [comment]
One thing to note though, I’m not sure if we want to just pass in the DatasetEvent object—this is an ORM model object and should generally be considered internal. Not sure how best to do this; maybe we need a separate data class similar to how Dataset is exposed to the user instead of the internal DatasetModel.

**Comment by SCrocky** [comment]
Hello @uranusjr,

I am highly interested in this feature as I'm working on a system for synchronizing datasets across multiple Airflow instances at my day job. 

But of course, I don't wish to intrude. @jrderuiter are you still working on this? If so can I offer any assistance?

**Comment by SCrocky** [comment]
@uranusjr Of course! I'll take a crack at it.

**Comment by uranusjr** [comment]
We’ll need to do something here for partitions. Maybe not exactly what’s proposed here, but something. (And the proposed addition is the most backward-compatible way that I can think of.)

cc @dstandish

**Comment by dstandish** [comment]
I am closing this one.  It is replaced by https://github.com/orgs/apache/projects/542/views/2?pane=issue&itemId=139253269&issue=apache%7Cairflow%7C58290

**Comment by uranusjr** [comment]
`on_asset_event_emitted` has been added in #61718 and will be a part of 2.3.
